### PR TITLE
feat: add details to api exception

### PR DESCRIPTION
**Pull request by ashish-s**
# Feature #

Add a new property named details to ApiException and RestException. Details is returned by selective apis to populate extra information about errors.
The Studio V2 flow api will be returning extra details with flow validation errors.

### Checklist
- [x] I acknowledge that all my contributions will be made under the project's license
- [x] I have made a material change to the repo (functionality, testing, spelling, grammar)
- [x] I have read the [Contribution Guidelines](CONTRIBUTING.md) and my PR follows them
- [x] I have titled the PR appropriately
- [x] I have updated my branch with the master branch
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation about the functionality in the appropriate .md file
- [x] I have added inline documentation to the code I modified

If you have questions, please file a [support ticket](https://twilio.com/help/contact), or create a GitHub Issue in this repository.

### CODE REVIEW

File: src/Twilio/Exceptions/ApiException.cs [hunk 1]
```diff
@@ -22,6 +23,11 @@ public class ApiException : TwilioException
         /// </summary>
         public string MoreInfo { get; }
 
+        /// <summary>
+        /// Any more details about the error, if any was provided
```
Comment (childish-sambino) [inline src/Twilio/Exceptions/ApiException.cs:27]: `s/was/were`

File: src/Twilio/Exceptions/ApiException.cs [hunk 2]
```diff
@@ -22,6 +23,11 @@ public class ApiException : TwilioException
...
+        /// </summary>
+        public Dictionary<string,object> Details { get; }
```
Comment (childish-sambino) [inline src/Twilio/Exceptions/ApiException.cs:29]: nit: spacing `string, object`

File: src/Twilio/Exceptions/ApiException.cs [hunk 3]
```diff
@@ -42,19 +48,21 @@ public class ApiException : TwilioException
...
         /// <param name="message">Error message</param>
         /// <param name="moreInfo">More info if provided</param>
+        /// <param name="details">Details if provided</param>
         /// <param name="exception">Original exception</param>
         public ApiException(
             int code,
             int status,
             string message,
             string moreInfo,
+            Dictionary<string,object> details = null,
```
Comment (childish-sambino) [inline src/Twilio/Exceptions/ApiException.cs:58]: nit: spacing `string, object`

File: test/Twilio.Test/Clients/TwilioRestClientTest.cs [hunk 4]
```diff
@@ -62,5 +63,33 @@ public void TestNotOkResponse()
...
+                                    ""message"": ""Bad request"",
+                                    ""more_info"": ""https://www.twilio.com/docs/errors/20001"",
+                                    ""status"": 400,
+                                    ""details"": {
+                                        ""foo"":""bar""
+                                    }}";
+            client.MakeRequest(Arg.Any<Request>()).Returns(new Response(HttpStatusCode.BadRequest, jsonResponse));
+            try {
+                Request request =  new Request(HttpMethod.Get,"https://www.contoso.com");
+                TwilioRestClient twilioClient = new TwilioRestClient("foo","bar",null,null,client);
```
Comment (childish-sambino) [inline test/Twilio.Test/Clients/TwilioRestClientTest.cs:81]: nit: arg spacing

File: test/Twilio.Test/Clients/TwilioRestClientTest.cs [hunk 5]
```diff
@@ -62,5 +63,33 @@ public void TestNotOkResponse()
...
+                twilioClient.Request(request);
+                Assert.Fail("Should have failed");
+            } catch (ApiException e) {
+                Assert.AreEqual("Bad request", e.Message);
+                Assert.AreEqual(20001, e.Code);
+                Assert.AreEqual("https://www.twilio.com/docs/errors/20001", e.MoreInfo);
+                Assert.AreEqual(400, e.Status);
+                var expectedDetails = new Dictionary<string,object>();
```
Comment (childish-sambino) [inline test/Twilio.Test/Clients/TwilioRestClientTest.cs:89]: nit: spacing `string, object`

File: test/Twilio.Test/Clients/TwilioRestClientTest.cs [hunk 6]
```diff
@@ -62,5 +63,33 @@ public void TestNotOkResponse()
...
+        [Test]
+        public void TestBadResponseWithDetails()
+        {
+            string jsonResponse = @"{
+                                    ""code"": 20001,
+                                    ""message"": ""Bad request"",
+                                    ""more_info"": ""https://www.twilio.com/docs/errors/20001"",
+                                    ""status"": 400,
+                                    ""details"": {
+                                        ""foo"":""bar""
```
Comment (childish-sambino) [inline test/Twilio.Test/Clients/TwilioRestClientTest.cs:76]: nit: spacing `: ""bar""`

File: test/Twilio.Test/Clients/TwilioRestClientTest.cs [hunk 7]
```diff
@@ -62,5 +63,33 @@ public void TestNotOkResponse()
...
+                TwilioRestClient twilioClient = new TwilioRestClient("foo","bar",null,null,client);
+                twilioClient.Request(request);
+                Assert.Fail("Should have failed");
+            } catch (ApiException e) {
+                Assert.AreEqual("Bad request", e.Message);
+                Assert.AreEqual(20001, e.Code);
+                Assert.AreEqual("https://www.twilio.com/docs/errors/20001", e.MoreInfo);
+                Assert.AreEqual(400, e.Status);
+                var expectedDetails = new Dictionary<string,object>();
+                expectedDetails.Add("foo","bar");
```
Comment (childish-sambino) [inline test/Twilio.Test/Clients/TwilioRestClientTest.cs:90]: nit: spacing

**Comment by ashish-s** [comment]
@childish-sambino fixed the formatting issues.

**Pull request merged**